tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,594 @@
+               d3.select(parent.$.tooltip)
+                 .classed('hidden', false);
+             })
+         .on('mouseout',
+             function() {
+               d3.select(this)
+                 .classed('cell-hover', false)
+                 .style('opacity', 1.0);
+              d3.select(parent.$.tooltip).classed('hidden', true);
+             });
+    cards.exit().remove();
+  },
+  /**
+   * Draw the UI of chip to chip links.
+   */
+   drawLinks: function(linkData: Array<podviewer.proto.ChannelInfo>) {
+    let parent = this;
+    if (!linkData || linkData.length == 0 || !this._gLink) {
+      return;
+    }
+
+    // Handle links;
+    let links = this._gLink.selectAll('.link').data(linkData);
+
+    // attach the arrow from defs
+    links.enter()
+         .append('svg:path')
+         .attr('id', (d) => 'cid' + d.channelId)
+         .attr('stroke-width', 2)
+         .attr('stroke', 'red')
+         .attr('fill', 'none')
+         .attr('marker-end', 'url(#arrow)')
+         .style('visibility', 'hidden')
+         .merge(links)
+         .attr('d', (d) => parent.linkToPath(d));
+
+    // Handle deleted links.
+    links.exit().remove();
+    this._selectedChannelIdChanged(this.selectedChannelId);
+  },
+  /**
+   * Given the global core id, returns the (x, y) coordinates in the topology
+   * graph.
+   */
+  coreIdToPos: function(id: number): Position {
+    let p = this;
+    const chipId = Math.floor(id / 2);
+    const nodeId = id & 1;
+    const xDim = chipId % p._xDimension;
+    const yDim = Math.floor(chipId / p._xDimension);
+    const x =
+        p.getNodeXLoc(
+            Math.floor(xDim / p._hostXStride), xDim % p._hostXStride, nodeId) +
+                CHIP_GRID_SIZE / NODES_PER_CHIP / 2;
+    const y = p.getChipYLoc(
+                  Math.floor(yDim / HOST_Y_STRIDE), yDim % HOST_Y_STRIDE) +
+                  CHIP_GRID_SIZE / 2;
+    return {x: x, y: y};
+  },
+  /**
+   * Returns the svg path given the src and dst core and node id.
+   * @return Path in svg format.
+   */
+  linkToPath: function(link: podviewer.proto.ChannelInfo): string {
+    let p = this;
+    const src = p.coreIdToPos(link.srcCoreId);
+    const dst = p.coreIdToPos(link.dstCoreId);
+    const path = 'M ' + src.x + ' ' + src.y + 'L ' + dst.x + ' ' + dst.y;
+    return path;
+  },
+  /**
+   * Returns the text to visualize in the tool tips.
+   * @return String to render in tool tips.
+   */
+  _getToolTipText: function(data: TopoData): string {
+    let parent = this;
+    let res = 'pos: (' + data.ydim + ',' + data.xdim + ')\n';
+    res += 'host: ' + data.host + '\n';
+    res += 'chip id: ' + data.cid + '\n';
+    res += 'node id: ' + data.nid + '\n';
+    res += 'replica id: ' + data.rid + '\n';
+    if (parent.selectedMetricIdx >= 0) {
+      res += parent.metrics[parent.selectedMetricIdx].label + ' spends ' +
+          data.value.toFixed(2) + '&mu;s in total, ';
+      const pcnt = data.value / data.total * 100;
```

Comment:
Optional: there is a cool and new shiny way of doing this:
```
const nf = new Intl.NumberFormat(navigator.language, {style: 'percent', minimumFractionDigits: 2})
nf.format(.3223211) // results in "32.23%"
nf.format(.3223911) // results in "32.24%"
```

Reply:
Cool. Thanks.